### ISSUE TITLE
Adding holidays that are outside the training/test range actually improves the model???

### ISSUE THREAD

**Issue by FredericoCoelhoNunes** [facebook/prophet#1220]
I am forecasting sales, and passing a custom list of holidays to Prophet, using the holidays parameter. 
Let's say that my full DataFrame goes from February to August, but I actually have access to the holidays in the previous and following years. When I give Prophet the full list of holidays for the Feb-Aug period, I actually get worse results than when I give it the full list of holidays (including all years).

I have read the Prophet article and can't understand why this should be, or why holidays outside the date range should have any effect in the results.
Any thoughts? Thanks!

**Comment by bletham** [comment]
That's a bit surprising to hear. At fit time, only holidays that occur on dates in the training data are included in the model. That happens here: https://github.com/facebook/prophet/blob/master/python/fbprophet/forecaster.py#L496

Then at predict time, we drop any holidays that did not appear in the training data: https://github.com/facebook/prophet/blob/b151d6e612f32c213a85f3af74207d2d7a32aca2/python/fbprophet/forecaster.py#L476-L483

So I'm not sure what could be happening here. Is there example code you could share that produces the issue so I could look at it in more detail? If you can't share data, then could you do
```
print(m.train_holiday_names)
```
after fitting to see if indeed it is fitting with the same set of holidays in both models?

**Comment by FredericoCoelhoNunes** [comment]
Thanks for the suggestion. I remember at the time I even created a very simple notebook with a dummy function that I created, and a bunch of fake holidays A, and then added several different holidays B outside the time range and was able to obtain different results. I don't have the notebook anymore though, sorry!

**Comment by FredericoCoelhoNunes** [comment]
Hey, so I think I've managed to reproduce this problem again. This is all in a self-contained notebook which I would be more than glad to share. Until then, here is the Python code. It shouldn't take more than half a minute to run.

An overview of the steps:

* I generate and plot a test time series function, which we will try to predict;
* I add a fixed value every 50 days to the time series (also to the day before and after). These are the days where we will have holidays, and I did this so we can tell a difference in accuracy from adding valid holidays
* I test three cases: 
  * 1) without any hoiidays;
  * 2) with holidays every 50 days inside the valid time range (the valid holidays)
  * 3) with the valid holidays + several holidays outside the time range, which supposedly should not change the result.
* The final results for case 2) and 3) are different!

```
#!/usr/bin/env python
# coding: utf-8

# In[1]:


import pandas as pd
from fbprophet import Prophet
import numpy as np
import matplotlib.pyplot as plt
get_ipython().run_line_magic('matplotlib', 'inline')


# In[3]:


X = np.linspace(1, 100, 5000)

# The test time range for our problem
TIME_RANGE = pd.date_range(start='2020-01-01', periods = 5000)


# In[4]:


# Generating the signal
y = np.sqrt(5*X**2)*np.sin(X*0.755) + X**1.5

# Will have holidays every 50 days, so adding a fixed amount to the holidays andarea around
for i in range(50, 5000, 50):
    y[i-1] = y[i-1] + 100 
    y[i] = y[i] + 200
    y[i+1] = y[i+1] + 100 


# In[5]:


# Plotting
fig = plt.figure(figsize=(12,8))
plt.plot(X,y)


# In[6]:


def test_holidays(y, time_range, holidays=None):
    """
    Fits the above function, predicts, and prints the accuracy.
    Can receive a holidays dataframe.
    """
    df = pd.DataFrame({'ds': time_range, 'y': y})
    train_df = df.iloc[:4000, :].copy()
    test_df = df.iloc[4000:, :].copy()
    
    if holidays is not None:
        m = Prophet(holidays=holidays)
    else:
        m = Prophet()
    m.fit(train_df)
    
    predictions = m.predict(test_df)
    
    true_val = test_df.y.sum()
    predicted_val = predictions.yhat.sum()
    accuracy = 1 - abs(true_val - predicted_val) / true_val
    print("Accuracy: ", accuracy)
    
    return m, accuracy


# In[7]:


#Testing without any holidays
test_holidays(y, time_range=TIME_RANGE)


# In[9]:


# Making some random holidays inside the valid time range, every 50 days, all of type VALID
valid_holidays = list(range(50, 5000, 50))
valid_holidays = [TIME_RANGE[i] for i in valid_holidays]
valid_holidays_df = pd.DataFrame({
    'ds': valid_holidays,
    'holiday': "VALID"
})

valid_m, valid_accuracy = test_holidays(y, time_range=TIME_RANGE, holidays=valid_holidays_df)


# In[10]:


# Making some random holidays OUTSIDE the range
invalid_holidays_1 = pd.date_range(end='2000-12-01', periods = 200)  # making sure these periods are way far back/ahead in time
invalid_holidays_2 = pd.date_range(start='2100-10-01', periods = 200)
invalid_holidays = invalid_holidays_1.append(invalid_holidays_2)

# Just to make sure none of the invalid holidays intersects with our time range:
# This means they should make no change to the results
assert set(invalid_holidays).intersection(set(TIME_RANGE)) == set()

invalid_holidays_df = pd.DataFrame({
    'ds': invalid_holidays,
    'holiday': "INVALID"
})


# Concatenating with the valid holidays
all_holidays_df = pd.concat([invalid_holidays_df, valid_holidays_df])

invalid_m, invalid_accuracy = test_holidays(y, time_range=TIME_RANGE, holidays=all_holidays_df)


# In[12]:


# This assert should pass, in my opinion. The holidayd outside the time period shouldn't change the result.
assert valid_accuracy == invalid_accuracy


# In[13]:


# Prints
print(valid_m.train_holiday_names)
print('\n')
print(invalid_m.train_holiday_names)
```

**Comment by FredericoCoelhoNunes** [comment]
Also, printing the holiday names using the model that had the "invalid holidays" yields the following:

0    INVALID
1      VALID
dtype: object

Which seems to indicate the INVALID holidays are being used somehow in training, despite not being inside the train_df time range.

**Comment by bletham** [comment]
Thanks for the clear repro code! I was able to figure out what is happening.

My comment above that holidays not in the training data are dropped was incorrect. That's just for holidays specified with the built-in country holidays; if it is missing for a particular year, then it is left out, and will not be included for predict.

If a holiday is specified with `m = Prophet(holidays=holidays)`, then it will be included in the model. What happens is just that it will be included as a column of all 0s in the feature matrix. Theoretically, the l2 regularization on its coefficient should drive its coefficient to 0, and it will have no impact on the model.

As you note, in practice it can actually have impact, and does in this code example. So what's happening here: because this synthetic data was generated to have seasonality that doesn't match any of the built-in periods, the model fit is very poor. Poor model fit like this produces a likelihood surface that is very flat around the optimum. In these settings, small changes to the model can cause the default L-BFGS optimizer to terminate in different parts of the parameter space. So while this column of 0s doesn't have any effect on the model predictions, it does affect the model fitting just by changing the problem slightly. In a problem with better model fit the effect would probably not be noticeable, but with poor model fit we have seen high variance in termination with the L-BFGS optimizer. I'm quite sure this is what is happening here because if you switch the optimizer to Newton (which we have found performs more robustly in these bad-model-fit, flat-likelihood scenarios) the accuracy with or without invalid holiday is basically the same:
```
m.fit(df, algorithm='Newton')

# Without invalid holidays, accuracy = 0.8869927981527871
# With invalid holidays,    accuracy = 0.8869927981735595
```

So basically, adding this useless parameter into the model is impacting the model fitting, likely more than usual because the model fit on this time series is poor.

I think it would be good to actually drop holidays that don't show up in the training data. I started looking into doing this and it's going to be a little bit complicated because of the windows around holidays. If I create a holiday for Christmas and specify a lower window of -2, then under the hood it is creating a feature vector for Christmas, Christmas-1, and Christmas-2. If I have Christmas and Christmas-1 in my training data but not Christmas-2, then I would want to drop Christmas-2. Right now the logic for keeping/dropping holidays (for country holidays) is only at the level of the holiday and not at the level of offset. So we'd have to add some extra logic for keeping track of which offsets we want to use for each holiday.

In the meantime, I'd recommend removing holidays that don't appear in the training set since generally it won't be beneficial to model fitting to add unused parameters.

**Comment by FredericoCoelhoNunes** [comment]
Thanks for the in-depth explanation! I'm glad I could be helpful.

Cheers
